sinaloa/src/ec2_instance.rs

```diff
@@ -0,0 +1,317 @@
+        let aws_subnet = env::var("AWS_SUBNET").expect("Failed to read AWS_SUBNET environment variable.");
+        let aws_region = env::var("AWS_REGION").expect("Failed to read AWS_REGION environment variable.");
+        let aws_security_group_id = env::var("AWS_SECURITY_GROUP_ID").expect("Failed to read AWS_SECURITY_GROUP_ID environment variable.");
+
+        let subnet_option = format!("--subnet-id={:}", aws_subnet);
+
+        let ec2_result = Command::new("/usr/local/bin/aws")
+            .args(&["ec2", "run-instances", "--image-id", "ami-037dd1d3f98da4d50",
+                    "--instance-type", "m5.xlarge", "--enclave-options", "Enabled=true",
+                    "--region",  &aws_region, "--key-name", &aws_key_name,
+                    &subnet_option, "--security-group-ids", &aws_security_group_id,
+                    "--associate-public-ip-address",
+                    "--tag-specifications",  "ResourceType=instance,Tags=[{Key=Veracruz,Value=RootEnclave}]"])
+                        .output()
+            .map_err(|err| {
+                println!("EC2Instance::new failed to start ec2 instance:{:?}", err);
+                EC2Error::IOError(err)
+            })?;
+        if !ec2_result.status.success() {
+            let ec2_result_text = std::str::from_utf8(&ec2_result.stderr)
+                .map_err(|err| EC2Error::Utf8Error(err))?;
+            println!("ec2 result Stdout:{:}", ec2_result_text);
+            return Err(EC2Error::CLIError);
+        }
+
+        let ec2_result_text = std::str::from_utf8(&ec2_result.stdout)
+            .map_err(|err| EC2Error::Utf8Error(err))?;
+        let ec2_data: Value = serde_json::from_str(ec2_result_text)
+            .map_err(|err| EC2Error::SerdeJsonError(err))?;
+
+        let instance_id: &str = match &ec2_data["Instances"][0]["InstanceId"] {
+            Value::String(value) => value,
+            _ => return Err(EC2Error::IncorrectJson),
+        };
+        let private_ip: &str = match &ec2_data["Instances"][0]["PrivateIpAddress"] {
+            Value::String(value) => value,
+            _ => return Err(EC2Error::IncorrectJson),
+        };
+        println!("EC2Instance instance_id:{:?}", instance_id);
+        println!("EC2Instance private_ip:{:?}", private_ip);
+
+        std::thread::sleep(std::time::Duration::from_millis(30000));
+
+        let socket_port: u16 = 9090;
+
+        Ok(EC2Instance {
+            instance_id: instance_id.to_string(),
+            private_ip: private_ip.to_string(),
+            socket_port: socket_port,
+            socket_fd: None,
+        })
+    }
+
+    fn socket_connect(&mut self) -> Result<RawFd, EC2Error> {
+        let ip_addr: Vec<u8> = self.private_ip.split(".")
+            .map(|s| s.parse().expect("Parse error")).collect();
+        let inet_addr: InetAddr = InetAddr::new(IpAddr::new_v4(ip_addr[0], ip_addr[1], ip_addr[2], ip_addr[3]), self.socket_port);
```

Comment:
I think this code for taking a string, splitting it on '.', and then taking the four subcomponents has already been used once above.  Should this be factored out into a utility function in e.g. `veracruz-utils` that does proper error handling?

Reply:
Yes, that is what it does. I will look into adding a function inside `veracruz-utils`.